### ISSUE TITLE
[BUG] ID properties from associations are not brought through when type CUID

### ISSUE THREAD

**Issue by jbm1991** [cap-js/cds-typer#62]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Nature of Your Project

TypeScript

### Current Behavior

I have an entity `Example1` which has an association to `Example2` as shown below.

```ts
entity Example1 {
      demo: String(100);
      example2 : Association to Example2;
}

entity Example2 : cuid {
  description : String(100);
}
```

In my typescript file I attempt to reference the property `example2_ID` in my code as per the below, however this gets flagged as an error because the property does not exist according to the types. This is an error because this property does exist. Other key fields are brought through in other Entities, it just seems to be when CUID aspect is used that the ID field is not available in the types.

As a side point this issue didn't exist in V0.3.0 I guess because it used to extend type any rather than object and so now it is stricter?

### Expected Behavior

the _ID property from CUID aspect are available on parent entities.

### Steps To Reproduce

See description above.

### Environment

```markdown
- **OS**: Mac OS
- **Node**: 18
- **npm**: 10 (also 9)
- **cds-typer**: 0.7.0
- **cds**: 7.1.2
```

**Comment by daogrady** [comment]
Hi John,

thank you for reporting this gap! If I am not mistaken, then this behaviour would be fixed in https://github.com/cap-js/cds-typer/pull/58

The PR is currently pending as I would like to merge some other changes first. But if you could confirm that the PR fixes your issue then I would prioritise it.

Best,
Daniel

**Comment by jbm1991** [comment]
Hi Daniel, thanks for the quick response. That PR certainly looks like it would solve the problem as the example you give of bringing `b_y` through from `C`, would correlate with me wanting to bring `..._ID` through from `cuid aspect`. So yes I would say that it should work.

Thanks
John

**Comment by daogrady** [comment]
Hi John,

thanks for letting me know! I will make sure to get this PR into a release by tomorrow and let you know once it is available on NPM.

Best,
Daniel

**Comment by daogrady** [comment]
Hi John,

just letting you know that cds-typer 0.8.0 has just been released which includes the fix that addresses your issue. 

Best,
Daniel

**Comment by jbm1991** [comment]
Thanks for the quick turnaround Daniel! I'm pleased to report that has fixed my CUID related problem

**Comment by jbm1991** [comment]
However it has actually left a very similar bug in place strangely. It's basically the exact same problem however the key fields instead of coming from a CUID aspect are coming from another entity, so if anything it more closely aligns with the example in your PR. Is your PR limited to only 3 entities deep or does it recursively go down? The only difference I can see is that where it isn't working is 4 layers deep

Thanks
John

**Comment by jbm1991** [comment]
Hi Daniel,

An example schema which has this problem would be:

```
entity Example1 {
      demo: String(100);
      key example2 : Association to Example2;
}

entity Example2 {
  key example3: Association to Example3;
  key example4: Association to Example4;
}

entity Example3 {
  key ID : String(100);
  key example5: Association to Example5;
}

entity Example4 {
  key ID : String(100);
}

entity Example5 {
  key ID : String(100);
}
```

Essentially then when trying to access properties of Example1 such as `example2_example3_ID` or `example2_example3_example4_ID` which are valid, Typescript throws a property not available error. In the generated types file, the output of the above contains the associations, but not the key properties themselves:

```
export function _Example1Aspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
  return class Example1 extends Base {
        example2?: __.Association.to<Example2>;
        example2_example3?: __.Association.to<Example3>;
        example2_example4?: __.Association.to<Example4>;
  };
}
export class Example1 extends _._managedAspect(_Example1Aspect(__.Entity)) {}
export class Example1_ extends Array<Example1> {}
```

I would expect the above to look like:

```
export function _Example1Aspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
  return class Example1 extends Base {
        example2?: __.Association.to<Example2>;
        example2_example3?: __.Association.to<Example3>;
        **example2_example3_ID?: string;**
        **example2_example3_example5_ID?: string;**
        example2_example4?: __.Association.to<Example4>;
        **example2_example4_ID?: string;**
  };
}
export class Example1 extends _._managedAspect(_Example1Aspect(__.Entity)) {}
export class Example1_ extends Array<Example1> {}
```

Hopefully that makes sense!

Thanks
John

**Comment by daogrady** [comment]
Hi John,

thanks for providing the sample, I will look into it and get back to you.

Best,
Daniel

**Comment by daogrady** [comment]
Hi John,

just making sure I am understanding your expections correctly.
Your sample shows that you'd expect associations to be propagated transitively if they are keys:

```ts
  return class Example1 extends Base {
        example2?: __.Association.to<Example2>;
        example2_example3?: __.Association.to<Example3>;  // <- here
        example2_example3_ID?: string;
        example2_example3_example5_ID?: string;
        example2_example4?: __.Association.to<Example4>;  // <- and here
        example2_example4_ID?: string;
  };
```

From my understanding only the key fields from referenced entities will be bubbled through to referring entities. The CSN for your sample model supports this assumption:

```js
LinkedCSN {
  definitions: {
    Example1: entity {
      kind: 'entity',
      elements: {
        demo: string { type: 'cds.String', length: 100 },
        example2: Association {  ...  },
        example2_example3_ID: string { ... },
        example2_example3_example5_ID: string { ... },
        example2_example4_ID: string { ... }
      }
    },
```
(Note how there is no `example2_example3` or `example2_example4`)

Is this still in line with what you are looking for?

Best,
Daniel

**Comment by jbm1991** [comment]
Hi Daniel,

Thanks for the writeup and that makes sense to me. I assumed the Associations you tagged with 'here' in your example would come through, so I included them in my example, but as you point out that doesn't align with the CSN and so they should not come through. The important thing for me here is the key fields being propagated through from associations, not so much the associations themselves. Essentially if the type definition could have the properties as shown in your CSN example (1 association and 3 _ID fields) then that would be perfect.

Thanks
John

**Comment by daogrady** [comment]
Hi John,

thank you for the quick response! Just wanted to make sure we are pushing this issue in the right direction. You are absolutely correct about the `*_ID` fields. I will make sure they are propagated.

BEst,
Daniel

**Comment by daogrady** [comment]
Hi John,

I have just released 0.9.0 which should address your issue. If you find any more gaps, please let me know.

Best,
Daniel

**Comment by jbm1991** [comment]
Hi Daniel,

That works perfectly. Thank you very much for the quick resolution!

Thanks
John
